atactm.1
.\"****************************************************************
.\"Copyright (c) 1993 Bell Communications Research, Inc. (Bellcore)
.\"
.\"Permission to use, copy, modify, and distribute this material
.\"for any purpose and without fee is hereby granted, provided
.\"that the above copyright notice and this permission notice
.\"appear in all copies, and that the name of Bellcore not be
.\"used in advertising or publicity pertaining to this
.\"material without the specific, prior written permission
.\"of an authorized representative of Bellcore.  BELLCORE
.\"MAKES NO REPRESENTATIONS ABOUT THE ACCURACY OR SUITABILITY
.\"OF THIS MATERIAL FOR ANY PURPOSE.  IT IS PROVIDED "AS IS",
.\"WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
.\"****************************************************************
.\"	$Header: /users/source/archives/atac.vcs/RCS/atactm.1,v 3.9 1994/04/05 09:50:19 saul Exp $
.\"
.\"$Log: atactm.1,v $
.\"Revision 3.9  1994/04/05 09:50:19  saul
.\"FROM_KEYS
.\"
. \"Revision 3.9  94/04/05  09:50:19  saul
. \"Add DISTRIBUTION heading.
. \"
. \"Revision 3.8  94/04/04  10:08:10  jrh
. \"Add Release Copyright
. \"
. \"Revision 3.7  94/04/04  09:49:16  jrh
. \"Add Release Copyright
. \"
. \"Revision 3.6  93/10/28  11:59:00  ewk
. \"Change header to indicate release 3.3
. \"
. \"Revision 3.5  93/07/13  11:23:12  ewk
. \" Updated SEE ALSO to include atacLD; fixed typos.
. \"
. \"Revision 3.4  93/03/30  14:33:14  saul
. \"Change comments to point to base as point of contact.
. \"
. \"Revision 3.3  93/03/30  08:43:03  saul
. \"Edits from ewk.
. \"
. \"Revision 3.2  92/12/30  11:29:43  saul
. \"Add last change date
. \"
.\"Revision 3.1  92/12/03  09:00:34  saul
.\"typo
.\"
.\"Revision 3.0  92/11/06  07:46:18  saul
.\"propagate to version 3.0
.\"
.\"Revision 2.2  92/10/28  08:58:58  saul
.\"editorial changes
.\"
.\"Revision 2.1  92/09/30  10:47:01  saul
.\"*** empty log message ***
.\"
.\"-----------------------------------------------end of log
.\"
.TH ATACTM 1 "$Date: 1994/04/05 09:50:19 $" "ATAC release 3.3"
.UC 4
.SH NAME
atactm \- automatic test analysis for C programs \- trace manager
.SH SYNOPSIS
.B atactm
[
.B \-o
.IR new .trace 
] [ 
.I trace-file 
]
.br
.B atactm
.BI \-c \ cost
.BI \-n \ test-name
[
.B \-x
] [
.B \-o
.IR new .trace 
] [ 
.I trace-file 
]
.br
.B atactm \-d
.BI \-n \ test-name
[
.B \-x
] [
.B \-o
.IR new .trace 
] [ 
.I trace-file 
]
.br
.B atactm \-e
.BI \-n \ test-name
[
.B \-x
]
.B \-o
.IR new .trace 
[
.I trace-file 
]
.br
.B atactm \-{lL}
[
.BI \-n \ test-name
[
.B \-x
] ] [
.I trace-file 
]
.br
.B atactm \-r
.I new-test-name
[
.BI \-n \ test-name
[
.B \-x
] ] [
.B \-o
.IR new .trace 
] [ 
.I trace-file 
]
.SH DESCRIPTION
.I Atactm
manages trace files created by execution of a program compiled with
.IR atacCC (1).
A trace file contains coverage information for each test case executed
by a program.
.I Atactm
can list, rename, extract, or assign cost to selected test cases.
The options
.BR \-c ,
.BR \-d ,
.BR \-e ,
.BR \-l ,
.BR \-L ,
and 
.B \-r
select the operation to be performed and are mutually exclusive.
If any of these options except
.BR \-l " and " \-L
are specified the trace file is compressed.
If no option is specified, the only effect is to compress the trace file.
If no trace file is specified the value of the
.B ATAC_TRACE
environment variable is used, if set.
Otherwise 
.I  a.out.trace
is used.
If the trace file name is not a complete path and the trace file is
not found in the current directory, the directory named by the
.B ATAC_DIR
environment variable is searched.
The original trace file is over-written unless the
.B \-o
option is specified.
.SH OPTIONS
.TP
.BI \-c \ cost
Assign 
.I cost
to test cases specified with
.B \-n
.I test-name.
Test case cost is used by
.IR atacmin (1).
.TP
.B \-d
Delete test cases specified with
.B \-n
.I test-name.
.TP
.B \-e
Extract test cases specified with
.B \-n
.I test-name.
.TP
.B \-l
List test case names.
If output is to a terminal, names are formatted into columns.
If
.B \-n
.I test-name
is specified, only selected test cases are listed.
.TP
.B \-L
List test cases with time stamp, ATAC release, cost, and attributes.
Attributes may be 
.B f
\&\- frequency information missing or
.B c
\&\- test information corrupted.
If
.B \-n
.I test-name
is specified, only selected test cases are listed.
.TP
.BI \-n \ test-name
Select test cases matching
.I test-name
(See also
.BR \-x .)
Wild cards may be used in
.I test-name.
Wild cards are the same as those used for file names by sh(1) (*, ?, and [...]).
Quote marks may be needed to prevent
the shell from expanding the test name as file names.
Multiple test-cases may also be specified by repeating the
.B \-n
option.
.TP
.BI \-r \ new-test-name
Rename test cases specified with
.B \-n
.I test-name.
If
.B \-n
.I test-name
is omitted, all test cases are renamed.
Test case names are composed of alphanumeric characters and underscore.
A numeric suffix is appended to distinguish test cases with the same name.
.TP
.B \-x
Include only test cases
.I not
selected by
.B \-n
options.
.SH EXAMPLE
The following command lists test cases in
.I prog.trace
.PP
.nf
.in +.5i
.ft CB
atactm -l prog.trace
.in
.ft
.PP
The output looks like this:
.PP
.in
.ft
t.1  t.2  t.3
.in
.ft
.fi
.SH FILES
.IR a.out .trace
\- trace file.
.SH "SEE ALSO"
atac(1), atacCC(1), atacLD(1), ataclib(1), atacmin(1), sh(1).
.SH DISTRIBUTION
Copyright (c) 1993 Bell Communications Research, Inc. (Bellcore)
.PP
Permission to use, copy, modify, and distribute this material
for any purpose and without fee is hereby granted, provided
that the above copyright notice and this permission notice
appear in all copies, and that the name of Bellcore not be
used in advertising or publicity pertaining to this
material without the specific, prior written permission
of an authorized representative of Bellcore.  BELLCORE
MAKES NO REPRESENTATIONS ABOUT THE ACCURACY OR SUITABILITY
OF THIS MATERIAL FOR ANY PURPOSE.  IT IS PROVIDED "AS IS",
WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
.PP
Send comments or questions to atac@bellcore.com.


CHANGES
1998-08-22  Thomas E. Dickey  <dickey@clark.net>

	* configure.in: CF_CURSES_TERMCAP needs preset variables

	* atacterm.h: RCS_BASE

	* aclocal.m4: resync with my macros, including termcap/curses headers

1998-07-25  Thomas E. Dickey  <dickey@clark.net>

	* config.guess: 980823

1998-02-24  Thomas E. Dickey  <dickey@clark.net>

	* config.sub: 980823

1997-12-10  Thomas E. Dickey  <dickey@clark.net>

	* makefile.in:
	turn off suffixes (so GNU make doesn't convert install.sh to install).
	modify distclean rule to preserve config.guess and config.sub

1997-12-09  Thomas E. Dickey  <dickey@clark.net>

	* configure: 971211

	* config.guess, config.sub: RCS_BASE

	* aclocal.m4:
	add CF_CHECK_CACHE, CF_FUNC_TGETENT, CF_OUTCHAR, CF_SED_CONFIG_H

	* config_h.in:
	moved TARGET_xxx to configure.in, because td_config doesn't grok them.

	* portable.h: simplified ifdefs for atexit() and on_exit()

	* configure.in:
	use CF_CHECK_CACHE, CF_FUNC_TGETENT, CF_TYPE_OUTCHAR, CF_SED_CONFIG_H
	to simplify/modularize this.  also, moved some fixed #define's here,
	since td_config doesn't parse them properly.

1997-11-03  Thomas E. Dickey  <dickey@clark.net>

	* makefile.in: add EXTRA_CFLAGS

	* configure.in:
	use macros CF_GCC_WARNINGS, CF_STRIP_G_OPT, CF_STRIP_O_OPT and CF_CC_ANSI_CHECK

	* aclocal.m4: RCS_BASE

	* portable.h:
	quick hack to use autoconf symbol for saying we do prototypes

1997-11-01  Thomas E. Dickey  <dickey@clark.net>

	* configure.in:
	correct conflict between --enable-warnings and --enable-debug

1997-07-17  Thomas E. Dickey  <dickey@clark.net>

	* configure.in: check existence of 'rename()'

1997-05-11  Thomas E. Dickey  <dickey@clark.net>

	* portable.h: correct sign in LURSHIFT

1996-12-01  Thomas E. Dickey  <dickey@clark.net>

	* configure: RCS_BASE

	* configure.in: sys/wait.h, AC_TYPE_SIGNAL

	* makefile.in: missed a Makefile/makefile conversion

1996-11-11  Thomas E. Dickey  <dickey@clark.net>

	* makefile.in: show cd/make (for tracing log with vile)
	lowercase makefiles throughout

	* configure.in: generate all subdir-makefiles
	add test for yacc

1995-12-29  Thomas E. Dickey  <dickey@clark.net>

	* atac.lsm, makefile.in, configure.in: RCS_BASE

	* Version: document _this_ version

	* config_h.in: RCS_BASE

1995-12-28  Thomas E. Dickey  <dickey@clark.net>

	* README-TOO: RCS_BASE

	* version.h: make 'version_h' const to avoid unused-variable warnings

	* atacmin.1: corrected (missing comment-marker) for log-comments

	* atacLD.1: corrected (missing comment-marker) for log comments

1995-12-27  Thomas E. Dickey  <dickey@clark.net>

	* portable.h: define P_ macro

1994-09-26  Thomas E. Dickey  <dickey@clark.net>

	* Version: RCS_BASE

1994-09-13  Thomas E. Dickey  <dickey@clark.net>

	* install.sh: RCS_BASE

1994-08-08  saul

	* portable.h: FROM_KEYS

1994-05-21  Thomas E. Dickey  <dickey@clark.net>

	* mkdirs.sh: RCS_BASE

1994-04-18  saul

	* atac-tutorial.ps: FROM_KEYS

1994-04-05  saul

	* README, atactm.1, atacmin.1, ataclib.1, atacLD.1, atacCC.1, atac.1, Makefile:
	FROM_KEYS

1994-04-04  jrh

	* version.h: FROM_KEYS



atacysis/atactm.c
/****************************************************************
*Copyright (c) 1993 Bell Communications Research, Inc. (Bellcore)
*
*Permission to use, copy, modify, and distribute this material
*for any purpose and without fee is hereby granted, provided
*that the above copyright notice and this permission notice
*appear in all copies, and that the name of Bellcore not be
*used in advertising or publicity pertaining to this
*material without the specific, prior written permission
*of an authorized representative of Bellcore.  BELLCORE
*MAKES NO REPRESENTATIONS ABOUT THE ACCURACY OR SUITABILITY
*OF THIS MATERIAL FOR ANY PURPOSE.  IT IS PROVIDED "AS IS",
*WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
****************************************************************/
#ifdef MVS
#include <mvapts.h>
MODULEID(%M%,%J%/%D%/%T%)
#endif /* MVS */

static char atactm_c[] = 
	"$Header: /users/source/archives/atac.vcs/atacysis/RCS/atactm.c,v 3.6 1994/04/04 13:51:21 saul Exp $";
static char bellcoreCopyRight[] =
"Copyright (c) 1993 Bell Communications Research, Inc. (Bellcore)";

/*
*-----------------------------------------------$Log: atactm.c,v $
*-----------------------------------------------Revision 3.6  1994/04/04 13:51:21  saul
*-----------------------------------------------FROM_KEYS
*-----------------------------------------------
*Revision 3.6  94/04/04  13:51:21  saul
*Fix binary copyright.
*
*Revision 3.5  94/04/04  10:24:45  jrh
*Add Release Copyright
*
*Revision 3.4  93/08/11  14:39:58  saul
*VMS #include rejected by some cpp's
*
*Revision 3.3  93/08/04  15:52:29  ewk
*Added MVS and solaris support.  Squelched some ANSI warnings.
*
actm.c
*Revision 3.2  93/07/09  12:04:48  saul
*Change stat.h include for VMS
*
*Revision 3.1  92/12/30  07:52:35  saul
*RENAME_SUPPORTED ==> RENAME_SUPPORT error
*
*Revision 3.0  92/11/06  07:47:02  saul
*propagate to version 3.0
*
*Revision 2.10  92/11/02  15:47:45  saul
*remove unused errno definition
*
*Revision 2.9  92/11/02  11:42:34  saul
*use portable.h to deterimine use of rename()
*
*Revision 2.8  92/10/30  09:53:47  saul
*include portable.h
*
*Revision 2.7  92/10/29  10:59:36  saul
*Bad free()'s sometimes caused core dump on atactm -r.  Also memory leak.
*
*Revision 2.6  92/10/28  09:02:13  saul
*change #includes for portability
*
*Revision 2.5  92/10/01  13:29:02  saul
*Bug affecting -d and -e.  Index only should affect -l and -L only.
*
*Revision 2.4  92/09/30  11:45:59  saul
*Use ATAC_DIR & ATAC_TRACE.  Get rid of -R option.
*
*Revision 2.3  92/09/22  15:38:43  saul
*Trace compression.
*Overwrite temp trace file
*
*Revision 2.2  92/09/08  10:12:23  saul
*changed trace format and data structures
*
*Revision 2.1  92/09/08  09:59:58  saul
*Purdue trace management
*
*-----------------------------------------------end of log
*/
#include <stdio.h>
#include <ctype.h>
#include "portable.h"
#include "ramfile.h"
#ifdef vms
#include <stat.h>
#else
#ifndef MVS
#include <sys/stat.h>	/* for portability must be below "ramfile.h" */
#endif /* MVS */
#endif /* vms */
#include "man.h"

/* forward declarations */
#ifndef RENAME_SUPPORT
static int rename();
#endif /* RENAME_SUPPORT */
static int exists();
int main();
static char *prefixFileName();
tablestype *tableptr();
static void usage();

#define CHECK_MALLOC(p) if((p)==NULL)memoryError("Out of memory"),exit(1)

#ifndef MVS
#define DEFAULT_TRACEFILE "a.out.trace"
#endif /* MVS */

extern void init();

int iTestCase;

static tablestype	tables;

static void
usage(cmd)
char	*cmd;
{
fprintf(stderr,"Usage:\n");
fprintf(stderr,"%s [-o new.trace] [file.trace]\n", cmd);
fprintf(stderr,"%s -c cost -n testname [-x] [-o new.trace] [file.trace]\n", cmd);
fprintf(stderr,"%s -d -n testnames [-x] [-o new.trace] [file.trace]\n", cmd);
fprintf(stderr,"%s -e -n testnames [-x] -o new.trace [file.trace]\n", cmd);
fprintf(stderr,"%s -l [-n testnames [-x]] [file.trace]\n", cmd);
fprintf(stderr,"%s -L [-n testnames [-x]] [file.trace]\n", cmd);
fprintf(stderr,"%s -r newname [-n testnames [-x]] [-o new.trace] [file.trace]\n", cmd);
fprintf(stderr,"\t-c assign cost to named test case\n");
fprintf(stderr,"\t-d delete named test cases\n");
fprintf(stderr,"\t-e extract named test cases into new trace file\n");
fprintf(stderr,"\t-l list test cases\n");
fprintf(stderr,"\t-L list test cases with time, cost, version, etc.\n");
fprintf(stderr,"\t-r rename test case\n");
fprintf(stderr,"\t-R rename multiple test cases\n");
fprintf(stderr,"\t-n test names (wild cards ?, * [chars])\n");
fprintf(stderr,"\t-o output trace file; defaults to input trace\n");
fprintf(stderr,"\t-x excluded tests specified with -n\n");
}


tablestype *
tableptr()
{
    return &tables;
}

/*
* prefixFileName: Allocate a variant filename string by prefixing the
* non-directory part of filename with the string prefix.
*/
static char *
prefixFileName(filename, prefix)
char	*filename;
char	*prefix;
{
    char	*f;
    char	*p;
    char	*new;
    int		fLen;
    int		pLen;

    /*
    * Allocate new string.
    */
    fLen = strlen(filename);
    pLen = strlen(prefix);
    new = (char *)malloc(fLen + pLen + 1);
    CHECK_MALLOC(new);

    /*
     * Copy filename to new from right to left (starting with '\0') until '/'
     * or no more.
     */
    new += fLen + pLen;
    for (f = filename + fLen; f >= filename; --f) {
	if (*f == '/') break;
	*new-- = *f;
    }

    /*
    * Copy prefix to new from right to left.
    */
    for (p = prefix + pLen - 1; p >= prefix; --p) {
	*new-- = *p;
    }

    /*
    * Copy rest of filename to new.
    */
    for (;f >= filename; --f) {
	*new-- = *f;
    }

    return new + 1;
}

int
main(argc, argv)
int	argc;
char	*argv[];
{
#ifndef MVS
    char	*traceFile = NULL;
    char	*traceDir = NULL;
#endif /* MVS */
    char	*tracePath;
    struct cfile	*cf;
    char	*outfile;
    int		i;
    char	*p;
    int		function;
    int		cost = 0;
    char	*newName = NULL;
    char	*names;
    int		nMatch;
    int		deselect;
    char	*getenv();
    int		indexOnly;

    function = ' ';
    names = NULL;
    outfile = NULL;
    deselect = 0;
    indexOnly = 0;

    /*
    * Collect options.
    */
    for (i = 1; i < argc; ++i) {
	p = argv[i];
	if (*p != '-') break;
	while (*++p) switch(*p)
	{
	case 'c':
	    if (function != ' ') {
		fprintf(stderr, "conflicting options: -%c -%c\n", function, *p);
		usage(argv[0]);
		exit(1);
	    }
	    function = *p;
	    if (*++p == '\0') {
		if (++i == argc) {
		    fprintf(stderr, "argument missing for -%c\n", *--p);
		    usage(argv[0]);
		    exit(1);
		}
		p = argv[i];
	    }
	    cost = atoi(p);
	    for (;*p; ++p) {
		if (*p < '0' || *p > '9') {
		    fprintf(stderr, "%s: invalid cost\n", argv[i]);
		    usage(argv[0]);
		    exit(1);
		}
	    }
	    --p;		/* setup for next arg */
	    break;
	case 'd':
	case 'e':
	case 'l':
	case 'L':
	    if (function != ' ') {
		fprintf(stderr, "conflicting options: -%c -%c\n", function, *p);
		usage(argv[0]);
		exit(1);
	    }
	    function = *p;
	    break;
	case 'n':
	    if (*++p == '\0') {
		if (++i == argc) {
		    fprintf(stderr, "argument missing for -%c\n", *--p);
		    usage(argv[0]);
		    exit(1);
		}
		p = argv[i];
	    }
	    if (names == NULL) {
		names = (char *)malloc(strlen(p) + 1);
		CHECK_MALLOC(names);
		strcpy(names, p);
	    } else {
		names = (char *)realloc(names, strlen(names) + strlen(p) + 2);
		strcat(names, ",");
		strcat(names, p);
	    }
	    p += strlen(p) - 1;		/* setup for next arg */
	    break;
	case 'o':
	    if (outfile != NULL) {
		fprintf(stderr, "conflicting options: -%c -%c\n", *p, *p);
		usage(argv[0]);
		exit(1);
	    }
	    if (*++p == '\0') {
		if (++i == argc) {
		    fprintf(stderr, "argument missing for -%c\n", *--p);
		    usage(argv[0]);
		    exit(1);
		}
		p = argv[i];
	    }
	    outfile = p;
	    p += strlen(p) - 1;		/* setup for next arg */
	    break;
	case 'r':
	    if (function != ' ') {
		fprintf(stderr, "conflicting options: -%c -%c\n", function, *p);
		usage(argv[0]);
		exit(1);
	    }
	    function = *p;
	    if (*++p == '\0') {
		if (++i == argc) {
		    fprintf(stderr, "argument missing for -%c\n", *--p);
		    usage(argv[0]);
		    exit(1);
		}
		p = argv[i];
	    }
	    newName = p;
	    for (; *p; ++p) {
		if (!isalnum(*p) && *p != '_') {
		    fprintf(stderr, "%s: invalid\n", newName);
		    exit(1);
		}
	    }
	    --p;		/* setup for next arg */
	    break;
	case 'x':
	    deselect = 1;
	    break;
	default:
	    fprintf(stderr, "unknown option: -%c\n", *p);
	    usage(argv[0]);
	    exit(1);
	    break;
	}
    }

    /*
    * Collect trace files. (Only one allowed currently.)
    */
#ifndef MVS 
    if (argc == i) {
	traceFile = getenv("ATAC_TRACE");
	if (traceFile == NULL)
	    traceFile = DEFAULT_TRACEFILE;
	traceDir = getenv("ATAC_DIR");
    }
    else if (argc == i + 1) {
	traceFile = argv[i];
	if (exists(traceFile)) {
	    traceDir = NULL;
	} else {
	    traceDir = getenv("ATAC_DIR");
	}
    } else {
	fprintf(stderr, "%s: unexpected arguments\n", argv[i+1]);
	usage(argv[0]);
	exit(1);
    }
    if (traceDir && *traceFile != '/') {
	tracePath = (char *)malloc(strlen(traceFile) + strlen(traceDir) + 2);
	CHECK_MALLOC(tracePath);
	sprintf(tracePath, "%s/%s", traceDir, traceFile);
    } else {
	tracePath = (char *)malloc(strlen(traceFile) + 1);
	CHECK_MALLOC(tracePath);
	strcpy(tracePath, traceFile);
    }	
#else /* MVS */
    if (argc == i) {
	tracePath = "DD:ATACTRCE";
    } else if (argc == i + 1) {
	tracePath = argv[i];
    } else {
	fprintf(stderr, "%s: unexpected arguments\n", argv[i+1]);
	usage(argv[0]);
	exit(1);
    }
#endif /* MVS */
    
    /*
    * Validate options.
    */
    switch(function)
    {
    case 'c':
    case 'd':
    case 'e':
	if (names == NULL) {
	    fprintf(stderr, "%c: -n option required\n", function);
	    usage(argv[0]);
	    exit(1);
	}
	break;
    }

    switch(function)
    {
    case 'e':
	if (outfile == NULL) {
	    fprintf(stderr, "%c: -o option required\n", function);
	    usage(argv[0]);
	    exit(1);
	}
	break;
    case 'l':
    case 'L':
	indexOnly = 1;
	if (outfile) {
	    fprintf(stderr, "conflicting options: -%c -%c\n", function, 'o');
	    usage(argv[0]);
	    exit(1);
	}
	break;
    }

    /*
    * Read input trace file.
    */
    if (NULL == (cf = (struct cfile *)cf_openIn(tracePath))) {
	perror(tracePath);
	exit(1);
    }
    init(&tables);
    load_prev(tracePath, cf, &tables, indexOnly);
    cf_close(cf);

    /*
    * Do function.
    */
    switch(function)
    {
    case 'c':
	nMatch = 0;
	for (i = 0; i < tables.mems.iMemberCount; ++i) {
	    if (patMatch(names, tables.mems.members[i].pName, deselect)) {
		++nMatch;
		tables.mems.members[i].iCost = cost;
	    }
	}
	if (nMatch == 0) {
	    fprintf(stderr, "%s: no matches\n", names);
	    exit(1);
	}
	break;
    case 'd':
	nMatch = 0;
	for (i = 0; i < tables.mems.iMemberCount; ++i) {
	    if (patMatch(names, tables.mems.members[i].pName, deselect)) {
		++nMatch;
		tables.mems.members[i].iDelete = 1;
	    }
	}
	if (nMatch == 0) {
	    fprintf(stderr, "%s: no matches\n", names);
	    exit(1);
	}
	break;
    case 'e':
	nMatch = 0;
	for (i = 0; i < tables.mems.iMemberCount; ++i) {
	    if (patMatch(names, tables.mems.members[i].pName, deselect)) {
		++nMatch;
	    } else {
		tables.mems.members[i].iDelete = 1;
	    }
	}
	if (nMatch == 0) {
	    fprintf(stderr, "%s: no matches\n", names);
	    exit(1);
	}
	break;
    case 'l':
	nMatch = 0;
	for (i = 0; i < tables.mems.iMemberCount; ++i) {
	    if (patMatch(names, tables.mems.members[i].pName, deselect)) {
		columns(tables.mems.members[i].pName);
		++nMatch;
	    }
	}
	if (nMatch) {
	    columnsEnd();
	} else {
	    if (names != NULL) {
		fprintf(stderr, "%s: no matches\n", names);
	    } else {
		fprintf(stderr, "no tests\n");
	    }
	    exit(1);
	}
#ifndef MVS
	free(tracePath);
#endif /* MVS */
	return 0;
    case 'L':
	nMatch = 0;
	for (i = 0; i < tables.mems.iMemberCount; ++i) {
	    if (patMatch(names, tables.mems.members[i].pName, deselect)) {
		printf("%17.17s %5.5s %4d %c%c--- %s\n", 
		       tables.mems.members[i].pDate,
		       tables.mems.members[i].pVersion,
		       tables.mems.members[i].iCost,
		       tables.mems.members[i].iFreqFlag ? '-' : 'f',
		       tables.mems.members[i].iCorrupted ? 'c' : '-',
		       tables.mems.members[i].pName);
		++nMatch;
	    }
	}
	if (nMatch == 0 && names != NULL) {
	    fprintf(stderr, "%s: no matches\n", names);
	}
#ifndef MVS
	free(tracePath);
#endif /* MVS */
	return 0;
    case 'r':
	{
	    int		nMatch;
	    int		nextN;
	    char	**oldname;
	    int		i;

	    nMatch = 0;
	    nextN = testNo(&tables.mems, newName);
	    for (i = 0; i < tables.mems.iMemberCount; ++i) {
		oldname = &tables.mems.members[i].pName;
		if (patMatch(names, *oldname, deselect)) {
		    ++nMatch;
		    free(*oldname);
		    *oldname = (char *)malloc(strlen(newName) + 11);
		    CHECK_MALLOC(oldname);
		    sprintf(*oldname, "%s.%d", newName, nextN++);
		}
	    }
	    if (nMatch == 0) {
		fprintf(stderr, "%s: no matches\n", names);
		exit(1);
	    }
	}
	break;
    }

    if (names) free(names);
	
    /*
    * Create temporary output trace file name.
    */
    if (outfile == NULL) {
#ifdef MVS
	outfile = "DD:ATACTOUT";
#else /* MVS */
	outfile = prefixFileName(tracePath, "!");
    } else {
	free(tracePath);
	tracePath = NULL;
#endif
    }

    /*
    * Open output trace file.
    */
    if (NULL == (cf = (struct cfile *)cf_openOut(outfile))) {
	fprintf(stderr, "%s: can't open\n", outfile);
	exit(1);
    }

    /*
    * Write output trace.
    */
    dump(cf, &tables);
    cf_close(cf);

#ifndef MVS
    /*
    * Replace input trace file with temporary output trace file.
    */
    if (tracePath) {
	if (rename(outfile, tracePath) != 0) {
	    perror(tracePath);
	}
	unlink(outfile);
	free(outfile);
	free(tracePath);
    }
#endif /* MVS */

    return 0;
}

#ifndef MVS
static int
exists(path)
char *path;
{
    struct stat buf;

    if (stat(path, &buf) == 0) return 1;
    else return 0;
}
#endif /* MVS */

#ifndef RENAME_SUPPORT
static int
rename(from, to)
char	*from;
char	*to;
{
    char	*backup;

    backup = prefixFileName(to, "%");
    if (link(to, backup) != 0) return -1;
    if (unlink(to) != 0) return -1;
    if (link(from, to) != 0) {
	fprintf(stderr, "backup saved in %s\n", backup);
	return -1;
    }
    unlink(backup);
    free(backup);
    return unlink(from);
}
#endif
